portal-impl/src/com/liferay/portlet/usersadmin/util/UserReindexManager.java

```diff
@@ -0,0 +1,222 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portlet.usersadmin.util;
+
+import com.liferay.petra.lang.SafeClosable;
+import com.liferay.petra.string.StringBundler;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.messaging.proxy.ProxyModeThreadLocal;
+import com.liferay.portal.kernel.model.User;
+import com.liferay.portal.kernel.search.Indexer;
+import com.liferay.portal.kernel.search.IndexerRegistryUtil;
+import com.liferay.portal.kernel.search.SearchException;
+import com.liferay.portal.kernel.util.ListUtil;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.function.Function;
+
+/**
+ * @author Minhchau Dang
+ */
+public class UserReindexManager {
+
+	public static final UserReindexManager INSTANCE = new UserReindexManager();
+
+	public void reindex(final List<User> users) throws SearchException {
+		long[] userIds = ListUtil.toLongArray(users, User.USER_ID_ACCESSOR);
+
+		reindex(userIds);
+	}
+
+	public void reindex(long... userIds) throws SearchException {
+		List<UserIndexerRequest> newUserIndexerRequests = new ArrayList<>();
+
+		Indexer<User> indexer = IndexerRegistryUtil.nullSafeGetIndexer(
+			User.class);
+
+		for (long userId : userIds) {
+			UserIndexerRequest userIndexerRequest = new UserIndexerRequest(
+				userId, indexer);
+
+			UserIndexerRequest newUserIndexerRequest =
+				_queuedIndexerRequests.computeIfAbsent(
+					userIndexerRequest, Function.identity());
+
+			if (newUserIndexerRequest == userIndexerRequest) {
+				newUserIndexerRequests.add(userIndexerRequest);
+			}
+		}
+
+		if (ProxyModeThreadLocal.isForceSync()) {
+			_reindex(newUserIndexerRequests);
+		}
+		else {
+			_queuedIndexerRequestExecutorService.submit(
+				() -> _syncReindex(newUserIndexerRequests));
+		}
+	}
+
+	public void reindex(final User user) throws SearchException {
+		reindex(user.getUserId());
+	}
+
+	private void _reindex(List<UserIndexerRequest> userIndexerRequests) {
+		int i = 0;
+
+		for (UserIndexerRequest userIndexerRequest : userIndexerRequests) {
+			i++;
+
+			if (userIndexerRequest.isExecuting()) {
```

Comment:
Hey @holatuwol, thank you for the PR! I am even learning a good bit from it. I have just one question, though: shouldn't we check if the request is being executed inside the synchronized block, too? I'm not that skilled with concurrent code so I would like to understand it a bit.

Reply:
@brandizzi : As of right now, the only thing calling `userIndexerRequest.execute()` (which sets and unsets the executing flag) is within the synchronized block. So within the code for this pull request, there's no situation where `userIndexerRequest.isExecuting()` returns true while in the synchronized block.

If we wanted to be more future-proof so that other code can call `userIndexerRequest.execute()`, then it wouldn't hurt to add the additional check.

Reply:
If you want more details around how the synchronized block works in this case for managing concurrency, let me know on Slack, and we can work through it on a call.

Reply:
Thank you! That makes sense. I see that, since UserIndexerRequest is private, we are in total control of where the method is called, which makes everything safe. I just didn't catch this dynamic at first :)

I'm just waiting for CI results on a PR, but I guess it will pass. After that, I'll forward it to BChan ASAP. Thanks!